app/services/katello/event_daemon/runner.rb

```diff
@@ -0,0 +1,115 @@
+  module EventDaemon
+    class Runner
+      @services = {}
+      @cache = ActiveSupport::Cache::MemoryStore.new
+      @cache_keys = {}
+
+      class << self
+        def initialize
+          FileUtils.touch(lock_file)
+        end
+
+        def settings
+          SETTINGS[:katello][:event_daemon]
+        end
+
+        def pid
+          return unless pid_file && File.exist?(pid_file)
+
+          File.open(pid_file) { |f| f.read.to_i }
+        end
+
+        def pid_file
+          pid_dir.join('katello_event_daemon.pid')
+        end
+
+        def pid_dir
+          Rails.root.join('tmp', 'pids')
+        end
+
+        def lock_file
+          Rails.root.join('tmp', 'katello_event_daemon.lock')
+        end
+
+        def write_pid_file
+          return unless pid_file
+
+          FileUtils.mkdir_p(pid_dir)
+          File.open(pid_file, 'w') { |f| f.puts Process.pid }
+        end
+
+        def stop
+          return unless pid == Process.pid
+          Rails.logger.info "<<<<<<<<<<<<<<< STOPPING RUNNER >>>>>>>>>>>>>>>>>"
+          @monitor_thread.kill
+          @cache.clear
+          @services.values.each do |service_class|
+            service_class.close
+            Rails.logger.info "Clearing cache #{@cache_keys[service_class]}"
+            Rails.cache.delete(@cache_keys[service_class])
+          end
+          File.unlink(pid_file) if pid_file && File.exist?(pid_file)
+        end
+
+        def start
+          Rails.logger.info "((((((((((((((NOT RUNNABLE))))))))))))))" unless runnable?
+          return unless runnable?
+          lockfile = File.open(lock_file, 'r')
+          begin
+            lockfile.flock(File::LOCK_EX)
+            strtd = started?
+            Rails.logger.info("******************** ABORTING start; already started *************") if strtd
+            return if strtd # ensure it wasn't started while we waited for the lock
+            start_monitor_thread
+            write_pid_file
+
+            at_exit do
+              stop
+            end
+
+            Rails.logger.info("Katello event daemon started process=#{Process.pid}")
+          ensure
+            lockfile.flock(File::LOCK_UN)
+          end
+        end
+
+        def started?
+          Process.kill(0, pid)
+          true
+        rescue Errno::ESRCH, TypeError # process no longer exists or we had no PID cached
+          false
+        end
+
+        def start_monitor_thread
+          @monitor_thread = Thread.new do
+            Katello::EventDaemon::Monitor.new(@services.values).start
+          end
+        end
+
+        def runnable?
+          # avoid accessing the disk on each request
+          @cache.fetch('katello_event_daemon_runnable', expires_in: 1.minute) do
+            !started? && settings[:enabled] && !::Foreman.in_rake? && !Rails.env.test?
+          end
+        end
+
+        def register_service(name, klass)
+          @services[name] = klass
+          @cache_keys[klass] = "#{name.downcase}_status"
```

Comment:
you may elect to change the structure of the service registry so we can store more data about the services now and in the future

current example:
```
{
 :candlepin_events => ::Katello::CandlepinEventListener
}
```

possibility:

```
{
  :candlepin_events => {
     class: Katello::CandlepinEventListener
     cache_key: "candlepin_events_status"
  }
}
```

Reply:
Instead of storing the cache key with each service class we decided to use a single cache for all services, so I left `@services` as-is for now and eliminated `@cache_keys`.  (just posting this for the record here)